### PR TITLE
Consider newlines as spaces for space-sensitive parsing

### LINKED ISSUE #16594: incorrect determination of macro # of arguments
The macro invocation below fails with an error "wrong number of arguments".  Notice that the expression is spread over two lines.  Somehow the loader is seeing the text after `@einsum` as two (or more?) arguments to the einsum macro instead of just one.  This is version 0.4.5 (downloaded today)

The problem goes away if I concatenate the two lines into one.  Even more oddly, the problem goes away if I simply insert a space character after the plus sign that terminates the first line!

This problem is not special for the einsum macro; the same thing happens for other macros I tried.

``` julia
        @einsum ws.d_deltas[ind1,ind2,ind3,ind4] = ws.d_deltas_t1[ind1,ind2,ind3,ind4] +
         ws.d_deltas_t2[ind1,ind2,ind3,ind4] # (s,I,j,k)
```

### PR DESCRIPTION
**Pull request by TotalVerb**
Fix #16594 by treating newlines, tabs, etc. as spaces that make binary parsing take precedence over unary parsing in space-sensitive situations.

cc @stevengj

### CODE REVIEW

**Comment by stevengj** [comment]
Looks good to me, but @JeffBezanson should probably take a look, too.

**Comment by stevengj** [comment]
Note that #16594 occurred only for the special case of operators that are parsed as both unary and binary, so a more conservative version of this PR would just change the `unary-and-binary-ops` case.

**Comment by TotalVerb** [comment]
This issue manifested in a different way for colon. It's minor, but the new behaviour is better.

Before:
```julia
julia> [1 :
       2]
1×2 Array{Int64,2}:
 1  2
```

Now:
```julia
julia> [1 :
       2]
ERROR: syntax: line break in ":" expression
```

(this case is tested)

As for `~`, it is both a unary and binary operator, just a strange one, so #16594 does indeed apply.

Old
```julia
julia> :(@test 1 ~
               2)
ERROR: syntax: missing comma or ) in argument list
```

New
```julia
julia> :(@test 1 ~
               2)
:(@test @~(1,2))
```

**Comment by TotalVerb** [comment]
The initial version of this PR was incorrect, leading to test failures. I have corrected this PR.

**Comment by StefanKarpinski** [comment]
Failure is macOS-only. Probably unrelated.

**Comment by StefanKarpinski** [comment]
bump @JeffBezanson

**Comment by tkelman** [comment]
Jeff, please review, do we want this one or not?

**Comment by stevengj** [comment]
Bump.

**Comment by JeffBezanson** [comment]
This example from the linked issue seems like the biggest problem:

```
julia> [1 + 
       2 0]
1×2 Array{Int64,2}:
 3  0
```

(where the trailing space on the first line makes a difference). I wonder if it's possible to fix just that?
@stevengj Do you want to take a look at this? Otherwise I feel we're running out of time and we can't really deal with this.

**Comment by stevengj** [comment]
I fixed the merge conflict.

**Comment by stevengj** [comment]
The change seems pretty reasonable to me.   Breaking, but unlikely to affect real-world code.   It is 11th hour but this one has been hanging around for a decision for a while…

**Comment by StefanKarpinski** [comment]
Failure is real—does not pass parsing tests.

**Comment by TotalVerb** [comment]
I think the issue is that the introduced `space-before-next-token?` consumes the whitespace unlike the existing peek-char test. It can probably be resolved by testing the next character against the spaces without skipping them? I forget why I introduced the skip-ws now. I can take a quick look.

**Comment by TotalVerb** [comment]
Hold on, it looks like all the test failures are precisely the added tests (which are testing the issue that this fixes) which have bitrotted over time. The argument count for each of them is precisely off by one because the macro syntax has changed. The string one must have had  a few extra quotes once upon a time. Let's see if the new commit fixes it.

**Comment by stevengj** [comment]
The `:([x .+ y])` parsing had changed to a `vect` node, so I fixed that test (I hope) as well.

**Comment by TotalVerb** [comment]
One last test it looks like... This one because of ParseError binding deprecation.

**Comment by TotalVerb** [comment]
@StefanKarpinski  good to go now?

**Comment by StefanKarpinski** [comment]
macOS failure is #26725.

File: test/parse.jl [hunk 1]
```diff
@@ -226,6 +226,27 @@ end
...
+@test [1 +1] == [1 1]
+@test length(:(@x 1 +1 -1).args) == 5
+@test length(:(@x 1 + 1 -1).args) == 4
+@test length(:(@x 1 + 1 - 1).args) == 3
+@test length(:(@x 1 +
+                  1 -
+                  1).args) == 3
+@test length(:(@x 1 +
+                  1 +
+                  1).args) == 3
```
Comment (JeffBezanson) [inline test/parse.jl]: I would prefer these tests to check more than just the number of arguments.

Comment (stevengj) [reply test/parse.jl]: Yeah, just check Expr equality?

**Review by JeffBezanson — APPROVED** [review]

**Pull request merged**